src/rest-server/src/models/v2/job/k8s.js

```diff
@@ -615,6 +648,100 @@ const getSshInfo = async (frameworkName) => {
   throw createError('Not Found', 'NoJobSshInfoError', `SSH info of job ${frameworkName} is not found.`);
 };
 
+const generateExitDiagnostics = (diag) => {
+  const podCompletionStatus = extractPodCompletionStatus(diag);
+  if (podCompletionStatus === null) {
+    return diag;
+  }
+
+  const regex = /matched: (.*)/;
+  const matches = diag.match(regex);
+  const summmaryInfo = diag.substring(0, matches.index + 'matched:'.length);
+  return summmaryInfo + '\n' + yaml.safeDump(podCompletionStatus);
+};
+
+const extractPodCompletionStatus = (diag) => {
+  if (_.isEmpty(diag)) {
+    return null;
+  }
+  // The pattern for diagnostics is "podPattern unmatched: json" or "podPattern matched: json"
+  const regex = /matched: (.*)/;
+  const matches = diag.match(regex);
+  if (matches.length < 2) {
+    return null;
+  }
+  try {
+    return JSON.parse(matches[1]);
+  } catch (error) {
+    logger.warn('Get diagnostics info failed', error);
+  }
+  return null;
+};
+
+const extractRuntimeOutput = (diag) => {
+  if (_.isEmpty(diag)) {
+    return null;
+  }
+
+  const podCompletionStatus = extractPodCompletionStatus(diag);
+  if (podCompletionStatus === null) {
+    return null;
+  }
+
+  let res = null;
+  for (const container of podCompletionStatus.containers) {
+    if (container.code <= 0) {
+      continue;
+    }
+    const message = container.message;
+    if (message == null) {
+      continue;
+    }
+    const anchor1 = /\[PAI_RUNTIME_ERROR_START\]/;
+    const anchor2 = /\[PAI_RUNTIME_ERROR_END\]/;
+    const match1 = message.match(anchor1);
+    const match2 = message.match(anchor2);
+    if (match1 !== null && match2 !== null) {
+      const start = match1.index + match1[0].length;
+      const end = match2.index;
+      const output = message.substring(start, end).trim();
+      res = yaml.safeLoad(output);
+      break;
+    }
+  }
+  return res;
+};
+
+const extractLauncherOutput = (diag) => {
+  const anchor = /\[PAI_RUNTIME_ERROR_START\]/;
+  const match = diag.match(anchor);
+  if (match != null) {
+    return null;
+  }
+  return diag;
+};
+
```

Comment:
merge extractLauncherOutput with extractLauncherOutput, since they depend on each other and you just need to match once?

to be generateAppExitMessages?

Reply:
You mean merge extractLauncherOutput  and extractRuntimeOutput here?